modules/data.land/R/soilgrids_soc_extraction.R

```diff
@@ -0,0 +1,209 @@
+
+    if (is.null(site_info)) {
+    stop("Please provide a BETY DB site list containing at least the site id and PostGIS geometry\
+    as lon and lat")
+  }
+  
+  # prepare site info for extraction
+  internal_site_info <- data.frame(site_info$id, site_info$sitename, site_info$lat,site_info$lon)
+  #create a variable to store mean and quantile of organic carbon density (ocd) for each soil depth
+  ocdquant <- matrix(NA, nrow = 6, ncol = length(internal_site_info$site_info.lon) * 4) #row represents soil depth, col represents mean, 5%, 50% and 95%-quantile of ocd for all sites 
+  #create a variable for site ID
+  siteid <- matrix(rep(as.numeric(site_info$id),times=4*6), nrow=6,byrow=TRUE)
+  lonlat <- cbind(internal_site_info$site_info.lon, internal_site_info$site_info.lat)
+  base_data_url <- "/vsicurl?max_retry=30&retry_delay=60&list_dir=no&url=https://files.isric.org/soilgrids/latest/data/ocd/ocd_"
+  depths <- c("0-5cm", "5-15cm", "15-30cm", "30-60cm", "60-100cm", "100-200cm")
+  layer_thick <- c(0.05,0.10,0.15,0.30,0.40,1.00) # in unit m
+ 
+  # reproject locations to soilgrids projection
+  #Soilgrids data is using Homolosine projection https://www.isric.org/explore/soilgrids/faq-soilgrids 
+  p <- terra::vect(lonlat, crs = "+proj=longlat +datum=WGS84") # Users need to provide lon/lat
+  newcrs <- "+proj=igh +datum=WGS84 +no_defs +towgs84=0,0,0" 
+  p_reproj <- terra::project(p, newcrs) # Transform the point vector to data with Homolosine projection
+  
+  # setup progress bar
+  if (verbose) {
+    j <- 1
+    pb <- utils::txtProgressBar(min = 0, max = length(depths), char="*", width=70, style = 3)
+  }
+  
+  for (dep in seq_along(depths)) {
+    # setup virtual raster URLs for each layer
+    ocd_mean.url <- paste0(base_data_url,depths[dep],"_mean.vrt")
+    ocd_Q0.05.url <- paste0(base_data_url, depths[dep], "_Q0.05.vrt")
+    ocd_Q0.50.url <- paste0(base_data_url, depths[dep], "_Q0.5.vrt")
+    ocd_Q0.95.url <- paste0(base_data_url, depths[dep], "_Q0.95.vrt")
+    
+    # create virtual rasters && extract SOC values - the original unit is hg/m3
+    ocd_mean <- terra::extract(terra::rast(ocd_mean.url), p_reproj)
+    ocd_Q0.05_map <- terra::extract(terra::rast(ocd_Q0.05.url), p_reproj)
+    ocd_Q0.50_map <- terra::extract(terra::rast(ocd_Q0.50.url), p_reproj)
+    ocd_Q0.95_map <- terra::extract(terra::rast(ocd_Q0.95.url), p_reproj)
+    
+    #change the unit to more common kg/m3
+    ocd_mean_real <- ocd_mean[, -1] / 10
+    ocd_Q0.05_real <- ocd_Q0.05_map[, -1] / 10
+    ocd_Q0.50_real <- ocd_Q0.50_map[, -1] / 10
+    ocd_Q0.95_real <- ocd_Q0.95_map[, -1] / 10
+    
+    ocdquant[dep, ] <-c(ocd_mean_real,ocd_Q0.05_real,ocd_Q0.50_real,ocd_Q0.95_real)
+    ### Display progress to console
+    if (verbose) {
+      utils::setTxtProgressBar(pb, j)
+      j <- j+1
+      flush.console()}
+  
+  # cleanup interim results
+    rm(ocd_mean.url, ocd_Q0.05.url, ocd_Q0.50.url, ocd_Q0.95.url, 
```

Comment:
This deallocation of memory, and reallocation at the start of the loop, probably slows down your code